sc-memory/sc-memory/src/scs/scs_parser.cpp

```diff
@@ -413,44 +456,109 @@ ElementHandle Parser::ProcessIdentifier(std::string const & name)
+  ScType type = scs::TypeResolver::GetElementType(scsType);
+  type |= scs::TypeResolver::IsConst(name) ? ScType::Const : ScType::Var;
+  return AppendElement(name, type);
 }
 
-void Parser::ProcessTriple(ElementHandle const & source, ElementHandle const & connector, ElementHandle const & target)
+void Parser::ProcessTriple(
+    ElementHandle const & sourceHandle,
+    ElementHandle const & connectorHandle,
+    ElementHandle const & targetHandle)
 {
-  ParsedElement const & connectorEl = GetParsedElement(connector);
+  ParsedElement & connector = GetParsedElementRef(connectorHandle);
 
-  auto AddConnector = [this, &connectorEl](ElementHandle const & src, ElementHandle const & e, ElementHandle const & trg) {
-    ParsedElement const & srcEl = GetParsedElement(src);
-    std::string const & idtf = srcEl.GetIdtf();
-    if (connectorEl.GetType() == ScType::ConstPermPosArc && scs::TypeResolver::IsKeynodeType(idtf))
-    {
-      ParsedElement & targetEl = GetParsedElementRef(trg);
-      ScType const newType = targetEl.m_type | scs::TypeResolver::GetKeynodeType(idtf);
+  auto AddConnector =
+      [this, &connector](
+          ElementHandle const & sourceHdl, ElementHandle const & connectorHdl, ElementHandle const & targetHdl)
+  {
+    ParsedElement const & source = GetParsedElement(sourceHdl);
+    ParsedElement & target = GetParsedElementRef(targetHdl);
 
-      if (targetEl.m_type.CanExtendTo(newType))
-        targetEl.m_type = newType;
-      else
-        SC_THROW_EXCEPTION(utils::ExceptionParseError, "Can't merge types for element " + targetEl.GetIdtf());
+    if (source.IsElementType() && (connector.GetType() == ScType::ConstPermPosArc))
+    {
+      std::string const & sourceIdtf = source.GetIdtf();
+      ScType const sourceType = scs::TypeResolver::GetElementType(sourceIdtf);
+      ScType const targetType = target.m_type;
+      ScType const newTargetType = targetType | sourceType;
+
+      if (target.IsElementType() && !newTargetType.CanExtendTo(ScType::ConstNodeClass))
+        SC_THROW_EXCEPTION(
+            utils::ExceptionParseError,
+            "Can't extend type `" << std::string(targetType) << "` using type `" << std::string(sourceType)
+                                  << "` for specified sc-element, because sc-element `" << target.GetIdtf()
+                                  << "` is sc-element denoting type of sc-elements.");
+      else if (ScType(targetType.BitAnd(~(ScType::Const | ScType::Var))).CanExtendTo(sourceType))
+        target.m_type = newTargetType;
+      else if (!sourceType.CanExtendTo(targetType))
+        SC_THROW_EXCEPTION(
+            utils::ExceptionParseError,
+            "Can't extend type `" << std::string(targetType) << "` using type `" << std::string(sourceType)
+                                  << "` for specified sc-element, because `" << std::string(sourceType) 
+                                  << "` is not subtype of `" << std::string(targetType) << "`.");
```

Comment:
also this subtype is questionable
usually it is correct, but when I add some tuple node to sc_node sourceType(sc_node) is not subtype of targetType(ConstNodeTuple) but everything is fine
probably this part of message cannot cover all cases and should be removed?

Reply:
sc_node_tuple is subtype of sc_node

Reply:
it is
but I am saying that when I add tuple to sc_node it is expected that this line will throw an exception because source type is not subtype of target type(sc_node is not subtype of sc_node_tuple)
and this is why subtype mention is questionable